homeassistant/components/yeelight/config_flow.py

```diff
@@ -0,0 +1,250 @@
+"""Config flow for Yeelight integration."""
+import logging
+
+import voluptuous as vol
+import yeelight
+
+from homeassistant import config_entries, exceptions
+from homeassistant.const import CONF_DISCOVERY, CONF_IP_ADDRESS, CONF_TYPE
+from homeassistant.core import callback
+from homeassistant.helpers import device_registry as dr
+import homeassistant.helpers.config_validation as cv
+
+from . import (
+    CONF_DEVICE,
+    CONF_MODE_MUSIC,
+    CONF_MODEL,
+    CONF_NIGHTLIGHT_SWITCH,
+    CONF_NIGHTLIGHT_SWITCH_TYPE,
+    CONF_SAVE_ON_CHANGE,
+    CONF_TRANSITION,
+    DATA_CONFIG_ENTRIES,
+    DATA_SCANNER,
+    NIGHTLIGHT_SWITCH_TYPE_LIGHT,
+    _async_default_config,
+)
+from . import DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+TYPE_DISCOVERY = "Discovery"
+TYPE_MANUAL = "Manual"
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Yeelight."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Return the options flow."""
+        if config_entry.data[CONF_DISCOVERY]:
+            return DiscoveryOptionsFlowHandler(config_entry)
+        return ManualOptionsFlowHandler(config_entry)
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self._capabilities = None
+        self._ipaddr = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        if user_input is not None:
+            if user_input[CONF_TYPE] == TYPE_DISCOVERY:
+                return await self.async_step_discovery_setup()
+            return await self.async_step_manual()
+
+        for entry in self._async_current_entries():
+            if entry.data[CONF_DISCOVERY]:  # Only one discovery entry allowed
+                return await self.async_step_manual()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_TYPE): vol.In([TYPE_DISCOVERY, TYPE_MANUAL])}
+            ),
+        )
+
+    async def async_step_discovery_setup(self, user_input=None):
+        """Handle discovery setup."""
+        if user_input is not None:
+            # Check if there is at least one device
+            # Run 3 times as packets can get lost
+            for _ in range(3):
+                devices = await self.hass.async_add_executor_job(
+                    yeelight.discover_bulbs
+                )
+                if len(devices) > 0:
+                    return self.async_create_entry(
+                        title="Discovery", data={CONF_DISCOVERY: True},
+                    )
+            return self.async_abort(reason="no_devices_found")
+
+        return self.async_show_form(step_id="discovery_setup")
+
+    async def async_step_manual(self, user_input=None):
```

Comment:
Instead of having a manual step in the main flow, could we just always create one config entry with discovery turned on or off. Then make an options flow where we discover devices and let the user select from those devices to add?

Reply:
The manual step is here to support devices that can not be discovered, e.g. in a different sub-net. Merge everything into one entry will make things more complicated.

Reply:
We can allow user to specify a device that isn't discovered too. By using this suggestion we make the config entry with options coherent for all entries instead of having a special config entry for the discovery and then some other type for the manual entries.

Reply:
I know it feels strange to have a special config entry for discovery, but merge everything into one adds complexity to this integration. For example, we'll have to handle the case where some manually added devices are not ready when HA starts. We'll have to handle retries and log messages instead of just through an exception. What about we split every device into one config entry? That might be more intuitive for users.

Reply:
Devices can always go unavailable.

Reply:
Yeah that's true. So do you think it is better to have a separated config entry for each device (discovered and manually added) or one config entry for everything? I prefer separated entries. We can discover devices and and let the user select in the main flow.

Reply:
I'd suggest one entry with all devices. Discovering and adding manual devices could happen both in the main flow and in the options flow if we want. That will allow the user to search and add more devices after creating the initial entry with some devices.

Reply:
What is the benefit of having one entry with all devices? The functionalities of the two approaches are the same. I'm trying to implement this but it seems I'm just re-implementing adding and removing that is already handled by config flow. Also, what is the benefit of changing the current approach? Seems that it adds complexity to the integration.

Reply:
The inconsistency in the config entry data generates complexity at each place we need to use the config entry. We can already see those effects in the current code. And it will just keep growing when we add features that interact with the config entry.

We should make sure that all config entries from an integration represent the same thing. It should not be a mix where one entry is representing a discovery feature while the other entries represent specific devices.

If setup variation is required that variation should happen in the config flow. But the result structure of the config flow should look the same regardless of what decisions were made in the config flow.

Reply:
But even if we merge the config entries into one, we still need to handle the manual and discovery cases in the integration. There is going to be more complexity because now we need to handle adding and removing devices in options flow.

Reply:
We'll isolate the complexity to the config flow. That's the point. It will not effect using the config entry.

Reply:
Thanks for the explanation. Now I get it.

But I still think give each device, discovered or manually added, one config flow would be better. The config entries are going to have the same structure, and we don't need to implement adding and deleting in options flow because HA already does it. Also, importing is going to be easier to implement.

Reply:
That would be ok.